### PR TITLE
[WIP] Fixes #20409 - Correct RBAC for repo content removal

### PR DESCRIPTION
**Pull request by jturel**
We were not actually checking whether a user was able to modify a repo before allowing them to remove content from it. Now we do! Let me know if I've missed something such as including a more suitable message in the response.

Testing pending.

### CODE REVIEW

**Comment by jturel** [comment]
[test]

**Comment by jturel** [comment]
[test]

**Comment by jlsherrill** [comment]
This issue is actually present for other actions such as product update, repository update, etc...   Is there any way we could make this a bit more generic for use at least across a single controller, but maybe across other controllers?

**Comment by jturel** [comment]
On it!

**Comment by jturel** [comment]
@jlsherrill check the activation key controller changes I added. I think this could be a simple generic solution for each controller. If we want something that spans multiple controllers we will need, I think, to do some kind of metaprogramming with a common scheme for naming methods, or maintain some kind of mapping external to the controller correlating actions to permissions

File: app/controllers/katello/api/v2/activation_keys_controller.rb [hunk 1]
```diff
@@ -243,12 +243,6 @@ def index_relation
       activation_keys
     end
 
-    def find_activation_key
```
Comment (jturel) [inline app/controllers/katello/api/v2/activation_keys_controller.rb]: When using find_resource we should not have to use our own find_<resourcename> methods.

File: app/controllers/katello/api/v2/activation_keys_controller.rb [hunk 2]
```diff
@@ -6,11 +6,11 @@ class Api::V2::ActivationKeysController < Api::V2::ApiController
     before_action :find_environment, :only => [:index, :create, :update]
     before_action :find_optional_organization, :only => [:index, :create, :show]
     before_action :find_content_view, :only => [:index]
-    before_action :find_activation_key, :only => [:show, :update, :destroy, :available_releases, :copy, :product_content,
+    before_action :authorize
+    before_action :find_resource, :only => [:show, :update, :destroy, :available_releases, :copy, :product_content,
                                                   :available_host_collections, :add_host_collections, :remove_host_collections,
                                                   :content_override, :add_subscriptions, :remove_subscriptions,
                                                   :subscriptions]
-    before_action :authorize
```
Comment (jturel) [inline app/controllers/katello/api/v2/activation_keys_controller.rb]: I think this should actually be the first before_action - any reason not to?

File: app/controllers/katello/api/v2/activation_keys_controller.rb [hunk 3]
```diff
@@ -330,5 +324,22 @@ def activation_key_params
 
       key_params
     end
+
+    def parent_scope
```
Comment (jturel) [inline app/controllers/katello/api/v2/activation_keys_controller.rb]: The Foreman auth model enables authorization to look up into the parent scope(s). Things didn't quite work when doing that, so to keep with the current way of not looking into the parent scope we can override this method. Can move this to a parent controller, but this is for demonstration.

File: app/controllers/katello/api/v2/activation_keys_controller.rb [hunk 4]
```diff
@@ -330,5 +324,22 @@ def activation_key_params
...
+      {}
+    end
+
+    def action_permission
```
Comment (jturel) [inline app/controllers/katello/api/v2/activation_keys_controller.rb]: Apparently creating permissions via permission_generator.rb is not enough to map controller actions to permissions. Those are only for route level authorization. These permissions actually check the filters for the entity found by find_resource.

File: lib/katello/permission_creator.rb [hunk 5]
```diff
@@ -12,6 +12,7 @@ def define
       host_collections_permissions
       lifecycle_environment_permissions
       product_permissions
+      #repository_permissions
```
Comment (jturel) [inline lib/katello/permission_creator.rb]: I have these disabled at the moment, but we might consider making repository permissions distinct from product permissions. The foreman auth model does not (at least as it appears to me) support our desire to treat repos the same as products permission-wise.

**Comment by beav** [comment]
@jturel is this still WIP?

**Comment by jturel** [comment]
@beav it's not a current WIP but according to a chat I had with @bbuckingham this may be coming back in our radar soon. If that's the case I'd like to leave it open to help whosoever picks it up - perhaps me!

**Pull request closed without merging**